File: packages/gbnf-experiment/data/20260909T132655Z_2a16d72b/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
from .colorize import Color


def get_parent_stack_id(pointer, col) -> str:
    stack_ids = []
    parent = pointer.parent
    while parent:
        meta = parent.node.meta
        stack_ids.append(f"{meta.stack_id},{meta.path_id},{meta.step_id}")
        parent = parent.parent
    return col("<-", Color.GRAY).join(col(sid, Color.RED) for sid in stack_ids)

File: packages/gbnf-experiment/data/20260909T001426Z_1c28aa33/ported_implementation/src/grammar-graph/get-parent-stack-id.ts
import { Color } from './colorize';
import type { GraphPointer } from './graph-pointer';

export const getParentStackId = (
  pointer: GraphPointer,
  col: (value: string | number, color: string) => string,
): string => {
  const stackIds: string[] = [];
  let parent: GraphPointer | undefined = pointer.parent;
  while (parent) {
    stackIds.push(
      `${parent.node.meta.stackId},${parent.node.meta.pathId},${parent.node.meta.stepId}`,
    );
    parent = parent.parent;
  }
  const arrow = col('<-', Color.GRAY);
  return stackIds.map(stackId => col(stackId, Color.RED)).join(arrow);
};
